src/webgpu/api/operation/command_buffer/queries/occlusionQuery.spec.ts

```diff
@@ -0,0 +1,1076 @@
+  });
+
+g.test('occlusion_query,stencil')
+  .desc(
+    `
+      Test beginOcclusionQuery/endOcclusionQuery using stencil to occlude
+    `
+  )
+  .params(kQueryTestBaseParams)
+  .fn(async t => {
+    const { writeMask, renderMode, bufferOffset, querySetOffset } = t.params;
+    const kNumQueries = 30;
+    const resources = t.setup({
+      writeMask,
+      renderMode,
+      bufferOffset,
+      querySetOffset,
+      numQueries: kNumQueries,
+      depthStencilFormat: 'stencil8',
+    });
+    const {
+      vertexBuffer,
+      occlusionQuerySet,
+      renderTargetTexture,
+      depthStencilTexture,
+      pipeline,
+    } = resources;
+
+    const renderPassDescriptor: GPURenderPassDescriptor = {
+      colorAttachments: [
+        {
+          view: renderTargetTexture.createView(),
+          loadOp: 'clear',
+          storeOp: 'store',
+        },
+      ],
+      depthStencilAttachment: {
+        view: depthStencilTexture!.createView(),
+        stencilClearValue: 0,
+        stencilLoadOp: 'clear',
+        stencilStoreOp: 'store',
+      },
+      occlusionQuerySet,
+    };
+
+    await t.runQueryTest(
+      resources,
+      renderPassDescriptor,
+      (helper, queryIndex) => {
+        helper.setStencilReference(queryIndex % 2);
+        const queryHelper = helper.beginOcclusionQuery(queryIndex);
+        queryHelper.setPipeline(pipeline);
+        queryHelper.setVertexBuffer(vertexBuffer);
+        queryHelper.draw(3);
+        queryHelper.end();
+      },
+      (passed, queryIndex) => {
+        const expectPassed = queryIndex % 2 === 0;
+        t.expect(
+          !!passed === expectPassed,
+          `queryIndex: ${queryIndex}, was: ${!!passed}, expected: ${expectPassed}, ${name}`
+        );
+      }
+    );
+  });
+
+g.test('occlusion_query,sample_mask')
+  .desc(
+    `
+      Test beginOcclusionQuery/endOcclusionQuery using sample_mask to occlude
+
+      Set mask to 4 and draw quads in opposing corners of the texel.
+    `
+  )
+  .params(kQueryTestBaseParams.combine('sampleMask', [0, 2, 4, 6]))
+  .fn(async t => {
+    const { writeMask, renderMode, bufferOffset, querySetOffset, sampleMask } = t.params;
+    const kNumQueries = 30;
+    const sampleCount = 4;
+    const resources = t.setup({
+      writeMask,
+      renderMode,
+      bufferOffset,
+      querySetOffset,
+      numQueries: kNumQueries,
+      sampleCount,
+      sampleMask,
+    });
+    const { occlusionQuerySet, renderTargetTexture, pipeline } = resources;
+
+    const createQuad = (offset: number) => {
+      // prettier-ignore
+      return t.createVertexBuffer(new Float32Array([
+        offset + 0   , offset + 0   , 0,
+        offset + 0.25, offset + 0   , 0,
+        offset + 0   , offset + 0.25, 0,
+        offset + 0   , offset + 0.25, 0,
+        offset + 0.25, offset + 0   , 0,
+        offset + 0.25, offset + 0.25, 0,
+      ]));
+    };
+
+    const vertexBufferTL = createQuad(0);
+    const vertexBufferBR = createQuad(0.25);
```

Comment:
Wait.. am I crazy or should this be BL and TR? Or am I misunderstanding the coordinates?

Reply:
you're not crazy. Done